File: frontend/app/api/services/lock.ts
import { Department, Region } from "@/nextauth";
import http from "./httpInstance";

class LockService {
	async getLock(department?: Department, region?: Region) {
		try {
			const params = new URLSearchParams();
			if (department != null) params.append("department", String(department));
			if (region != null) params.append("region", String(region));
			const query = params.toString() ? `?${params}` : "";
			const response = await http.get(`/lock${query}`);
			return response.data as boolean;
		} catch (error) {
			console.error("Error while fetching Lock", error);
			throw error;
		}
	}

	async setLock(isLocked: boolean, department?: Department, region?: Region) {
		try {
			const params = new URLSearchParams({ is_locked: String(isLocked) });
			if (department != null) params.append("department", String(department));
			if (region != null) params.append("region", String(region));
			const response = await http.post(`/lock?${params}`);
			return response.data;
		} catch (error) {
			console.error("Error while setting the Lock", error);
			throw error;
		}
	}
}

const lockService = new LockService();
export default lockService;

File: backend/app/api/v1/routers/lock.py
from fastapi import APIRouter, Query

from app.core.deps import AllAuthenticated, DbSession, ManagementAndAbove
from app.models.enums import Department, Region
from app.schemas.common import MessageResponse
from app.services.dashboard_service import DashboardService

router = APIRouter(prefix="/lock", tags=["lock"], dependencies=[AllAuthenticated])


@router.get("", response_model=bool)
async def get_lock(
    db: DbSession,
    department: Department | None = Query(default=None),
    region: Region | None = Query(default=None),
) -> bool:
    return await DashboardService(db).get_lock(department, region)


@router.post("", response_model=MessageResponse, dependencies=[ManagementAndAbove])
async def set_lock(
    is_locked: bool,
    db: DbSession,
    department: Department | None = Query(default=None),
    region: Region | None = Query(default=None),
) -> MessageResponse:
    await DashboardService(db).set_lock(is_locked, department, region)
    return MessageResponse(message=f"Lock changed to {is_locked}")
